File: sore/app/lib/dijkstra.ts
import { GraphData } from './graph';

export function dijkstra(G: number[][][], start: number): [number[], number[]] {
  const n = G.length;
  const dist: number[] = Array(n).fill(Infinity);
  const parent: number[] = Array(n).fill(-1);
  dist[start] = 0;

  const pq: [number, number][] = [[0, start]];
  const visited = new Set<number>();

  while (pq.length > 0) {
    pq.sort((a, b) => a[0] - b[0]);
    const [d, u] = pq.shift()!;
    
    if (visited.has(u)) continue;
    visited.add(u);
    
    if (d > dist[u]) continue;

    for (const [v, weight] of G[u]) {
      if (!visited.has(v) && dist[u] + weight < dist[v]) {
        dist[v] = dist[u] + weight;
        parent[v] = u;
        pq.push([dist[v], v]);
      }
    }
  }

  return [parent, dist];
}

export function reconstruir(parent: number[], goal: number): number[] | null {
  if (goal === -1 || parent[goal] === -1) return null;

  const path: number[] = [];
  let current = goal;
  const visited = new Set<number>();
  
  while (current !== -1 && !visited.has(current)) {
    visited.add(current);
    path.push(current);
    current = parent[current];
  }
  
  if (current === -1) {
    return path.reverse();
  }
  
  return null; // Cycle detected
}

export function calcularRuta(graphData: GraphData, startLabel: string, endLabel?: string): {
  start: number;
  goal: number;
  cost: number;
  rutaIdx: number[];
  rutaLabels: string[];
} | null {
  const { G, labels, w2i, tipos_nodo } = graphData;

  if (!(startLabel in w2i)) return null;

  const start = w2i[startLabel];
  const [parent, dist] = dijkstra(G, start);

  let goal: number;
  if (endLabel) {
    if (!(endLabel in w2i)) return null;
    goal = w2i[endLabel];
    if (dist[goal] === Infinity) return null;
  } else {
    // Find nearest exit
    const exitsIdx = Object.entries(tipos_nodo)
      .filter(([n, t]) => t === "Salida" && n in w2i)
      .map(([n]) => w2i[n]);
    const candidatos = exitsIdx
      .map(e => [e, dist[e]] as [number, number])
      .filter(([e, d]) => d < Infinity);
    if (candidatos.length === 0) {
      // If no exits found, return null
      console.log('No exits found or reachable from start node');
      return null;
    }
    goal = candidatos.reduce((min, curr) => curr[1] < min[1] ? curr : min)[0];
  }

  const rutaIdx = reconstruir(parent, goal);
  if (!rutaIdx) return null;

  const cost = dist[goal];
  const rutaLabels = rutaIdx.map(i => labels[i]);

  return { start, goal, cost, rutaIdx, rutaLabels };
}

File: Hito-1/dijkstra.py
# Dijkstra algorithm for evacuation routes
import math
import heapq as hq

def dijkstra(G, start):
    n = len(G)
    dist = [math.inf] * n
    parent = [-1] * n
    dist[start] = 0
    
    pq = [(0, start)]
    
    while pq:
        d, u = hq.heappop(pq)
        if d > dist[u]:
            continue
        
        for v, weight in G[u]:
            if dist[u] + weight < dist[v]:
                dist[v] = dist[u] + weight
                parent[v] = u
                hq.heappush(pq, (dist[v], v))
    
    return parent, dist

def reconstruir(parent, goal):
    if goal is None or parent[goal] == -1:
        return None
    
    path = []
    current = goal
    while current != -1:
        path.append(current)
        current = parent[current]
    
    return path[::-1]

def calcular_ruta(G, labels, w2i, tipos_nodo, start_label, end_label=None):
    """Calculate route from start to end using Dijkstra"""
    if start_label not in w2i:
        return None, None, None, None
    
    start = w2i[start_label]
    parent, dist = dijkstra(G, start)
    
    if end_label:
        if end_label not in w2i:
            return None, None, None, None
        goal = w2i[end_label]
        if dist[goal] == math.inf:
            return None, None, None, None
        ruta_idx = reconstruir(parent, goal)
        cost = dist[goal]
    else:
        # Find nearest exit
        exits_idx = {w2i[n] for n,t in tipos_nodo.items() if t=="Salida" and n in w2i}
        goal = None
        if exits_idx:
            candidatos = [(e, dist[e]) for e in exits_idx if dist[e] < math.inf]
            if candidatos:
                goal = min(candidatos, key=lambda x: x[1])[0]
        if goal is None:
            return None, None, None, None
        ruta_idx = reconstruir(parent, goal)
        cost = dist[goal]
    
    return start, goal, cost, ruta_idx